File: crates/lvz-context/src/lang.rs
//! Language detection and per-language tree-sitter configuration.
//!
//! Each supported [`Lang`] knows its tree-sitter grammar and which definition nodes carry
//! an elidable `body` field — the knobs the skeletoniser (`super::skeleton`) needs to keep
//! signatures while dropping bodies.

use tree_sitter::Language as TsLanguage;

/// A source language Lavoisier can parse for skeletons and symbols.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Lang {
    Rust,
    Python,
    JavaScript,
    TypeScript,
}

/// How to elide bodies for a language.
pub(crate) struct LangSpec {
    /// Definition node kinds whose `body` field should be replaced by [`elision`](Self::elision).
    pub def_kinds: &'static [&'static str],
    /// Placeholder text the body is replaced with (kept syntactically suggestive, not valid).
    pub elision: &'static str,
    /// Named-definition node kinds that become nodes in the symbol-dependency graph.
    pub symbol_kinds: &'static [&'static str],
    /// Keep a leading docstring (a bare string statement as the body's first item) when eliding
    /// the rest of the body — §6.1 wants docstrings retained as high-signal context.
    /// True only for Python, whose `def` bodies conventionally open with a `"""…"""` docstring.
    pub keeps_docstring: bool,
    /// Leaf node kinds that count as a *reference* to a name when resolving symbol-dependency
    /// edges (`super::symbols`). Restricting to real identifier nodes (vs. raw substring search)
    /// is what makes a name in a string or comment stop creating a spurious edge.
    pub ref_ident_kinds: &'static [&'static str],
    /// Node kinds that introduce **local bindings** (parameters, `let`/variable declarations).
    /// The identifiers in their binding position (`pattern`/`name` field, else direct children)
    /// are locals, so they are excluded from references — this is the scope/shadowing fix: a local
    /// variable that happens to share a top-level symbol's name no longer links to that symbol.
    pub binder_kinds: &'static [&'static str],
}

impl Lang {
    /// Best-effort detection from a file path's extension.
    pub fn from_path(path: &str) -> Option<Lang> {
        let ext = path.rsplit('.').next()?.to_ascii_lowercase();
        Some(match ext.as_str() {
            "rs" => Lang::Rust,
            "py" | "pyi" => Lang::Python,
            "js" | "jsx" | "mjs" | "cjs" => Lang::JavaScript,
            "ts" | "tsx" | "mts" | "cts" => Lang::TypeScript,
            _ => return None,
        })
    }

    /// The tree-sitter grammar for this language.
    pub(crate) fn ts_language(self) -> TsLanguage {
        match self {
            Lang::Rust => tree_sitter_rust::LANGUAGE.into(),
            Lang::Python => tree_sitter_python::LANGUAGE.into(),
            Lang::JavaScript => tree_sitter_javascript::LANGUAGE.into(),
            Lang::TypeScript => tree_sitter_typescript::LANGUAGE_TYPESCRIPT.into(),
        }
    }

    pub(crate) fn spec(self) -> LangSpec {
        match self {
            Lang::Rust => LangSpec {
                def_kinds: &["function_item"],
                elision: "{ … }",
                symbol_kinds: &[
                    "function_item",
                    "struct_item",
                    "enum_item",
                    "trait_item",
                    "type_item",
                    "const_item",
                    "static_item",
                ],
                keeps_docstring: false,
                ref_ident_kinds: &["identifier", "type_identifier"],
                binder_kinds: &["parameter", "let_declaration", "closure_parameters"],
            },
            Lang::Python => LangSpec {
                def_kinds: &["function_definition"],
                elision: "...",
                symbol_kinds: &["function_definition", "class_definition"],
                keeps_docstring: true,
                ref_ident_kinds: &["identifier"],
                binder_kinds: &["parameters", "lambda_parameters"],
            },
            Lang::JavaScript => LangSpec {
                def_kinds: &[
                    "function_declaration",
                    "method_definition",
                    "function_expression",
                ],
                elision: "{ … }",
                symbol_kinds: &[
                    "function_declaration",
                    "method_definition",
                    "class_declaration",
                ],
                keeps_docstring: false,
                ref_ident_kinds: &["identifier"],
                binder_kinds: &["formal_parameters", "variable_declarator"],
            },
            Lang::TypeScript => LangSpec {
                def_kinds: &[
                    "function_declaration",
                    "method_definition",
                    "function_expression",
                ],
                elision: "{ … }",
                symbol_kinds: &[
                    "function_declaration",
                    "method_definition",
                    "class_declaration",
                    "interface_declaration",
                    "type_alias_declaration",
                ],
                keeps_docstring: false,
                ref_ident_kinds: &["identifier", "type_identifier"],
                binder_kinds: &[
                    "formal_parameters",
                    "variable_declarator",
                    "required_parameter",
                    "optional_parameter",
                ],
            },
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn detects_languages_by_extension() {
        assert_eq!(Lang::from_path("src/main.rs"), Some(Lang::Rust));
        assert_eq!(Lang::from_path("a/b/c.py"), Some(Lang::Python));
        assert_eq!(Lang::from_path("x.JSX"), Some(Lang::JavaScript));
        assert_eq!(Lang::from_path("x.tsx"), Some(Lang::TypeScript));
        assert_eq!(Lang::from_path("README.md"), None);
        assert_eq!(Lang::from_path("noext"), None);
    }
}

File: src/Lavoisier/Context/Lang.hs
{- | Language detection and per-language tree-sitter configuration (ports Rust
@lvz-context::lang@). Each supported 'Lang' knows which definition nodes carry an elidable
@body@ (for the skeletoniser) and which node kinds are symbols\/references\/binders (for the
symbol-dependency graph). The grammar pointer itself lives in "Lavoisier.Context.TreeSitter"
(it is FFI, not pure data); everything here is pure.
-}
module Lavoisier.Context.Lang (
    Lang (..),
    langFromPath,
    LangSpec (..),
    langSpec,
)
where

import Data.Char (toLower)
import Data.Text (Text)

-- | A source language Lavoisier can parse for skeletons and symbols.
data Lang
    = Rust
    | Python
    | JavaScript
    | TypeScript
    deriving stock (Bounded, Enum, Eq, Ord, Show)

-- | Best-effort detection from a file path's extension (case-insensitive).
langFromPath ∷ String → Maybe Lang
langFromPath path = case map toLower ext of
    "rs" → Just Rust
    "py" → Just Python
    "pyi" → Just Python
    "js" → Just JavaScript
    "jsx" → Just JavaScript
    "mjs" → Just JavaScript
    "cjs" → Just JavaScript
    "ts" → Just TypeScript
    "tsx" → Just TypeScript
    "mts" → Just TypeScript
    "cts" → Just TypeScript
    _ → Nothing
    where
        ext = reverse (takeWhile (/= '.') (reverse path))

-- | How to elide bodies and classify nodes for a language.
data LangSpec = LangSpec
    { defKinds ∷ [Text]
    -- ^ Definition node kinds whose @body@ field is replaced by 'elision'.
    , elision ∷ Text
    -- ^ Placeholder text a body is replaced with (syntactically suggestive, not valid).
    , symbolKinds ∷ [Text]
    -- ^ Named-definition node kinds that become nodes in the symbol-dependency graph.
    , keepsDocstring ∷ Bool
    -- ^ Keep a leading docstring when eliding the rest of a body (Python only).
    , refIdentKinds ∷ [Text]
    {- ^ Leaf node kinds that count as a /reference/ to a name (real identifiers only, so a
    name in a string or comment creates no spurious edge).
    -}
    , binderKinds ∷ [Text]
    {- ^ Node kinds that introduce local bindings (parameters, @let@\/variable declarations);
    identifiers in their binding position are locals, excluded from references.
    -}
    , importKinds ∷ [Text]
    {- ^ Import\/@use@ declaration kinds. Their raw text is mined for module-path segments, which
    /rank/ cross-file definitions of a name — never resolve them (see "Lavoisier.Context.Symbols").
    -}
    }
    deriving stock (Eq, Show)

-- | The elision\/symbol\/reference configuration for a language.
langSpec ∷ Lang → LangSpec
langSpec = \case
    Rust →
        LangSpec
            { defKinds = ["function_item"]
            , elision = "{ … }"
            , symbolKinds =
                [ "function_item"
                , "struct_item"
                , "enum_item"
                , "trait_item"
                , "type_item"
                , "const_item"
                , "static_item"
                ]
            , keepsDocstring = False
            , refIdentKinds = ["identifier", "type_identifier"]
            , binderKinds = ["parameter", "let_declaration", "closure_parameters"]
            , importKinds = ["use_declaration"]
            }
    Python →
        LangSpec
            { defKinds = ["function_definition"]
            , elision = "..."
            , symbolKinds = ["function_definition", "class_definition"]
            , keepsDocstring = True
            , refIdentKinds = ["identifier"]
            , binderKinds = ["parameters", "lambda_parameters"]
            , importKinds = ["import_statement", "import_from_statement"]
            }
    JavaScript →
        LangSpec
            { defKinds = ["function_declaration", "method_definition", "function_expression"]
            , elision = "{ … }"
            , symbolKinds = ["function_declaration", "method_definition", "class_declaration"]
            , keepsDocstring = False
            , refIdentKinds = ["identifier"]
            , binderKinds = ["formal_parameters", "variable_declarator"]
            , importKinds = ["import_statement"]
            }
    TypeScript →
        LangSpec
            { defKinds = ["function_declaration", "method_definition", "function_expression"]
            , elision = "{ … }"
            , symbolKinds =
                [ "function_declaration"
                , "method_definition"
                , "class_declaration"
                , "interface_declaration"
                , "type_alias_declaration"
                ]
            , keepsDocstring = False
            , refIdentKinds = ["identifier", "type_identifier"]
            , binderKinds =
                [ "formal_parameters"
                , "variable_declarator"
                , "required_parameter"
                , "optional_parameter"
                ]
            , importKinds = ["import_statement"]
            }
